src/condition/cmdi_detector.cpp

```diff
@@ -0,0 +1,263 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+#include <cstddef>
+#include <optional>
+#include <span>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "argument_retriever.hpp"
+#include "clock.hpp"
+#include "condition/base.hpp"
+#include "condition/cmdi_detector.hpp"
+#include "condition/shi_common.hpp"
+#include "condition/structured_condition.hpp"
+#include "ddwaf.h"
+#include "exception.hpp"
+#include "exclusion/common.hpp"
+#include "iterator.hpp"
+#include "log.hpp"
+#include "tokenizer/shell.hpp"
+#include "transformer/common/cow_string.hpp"
+#include "transformer/lowercase.hpp"
+#include "utils.hpp"
+
+using namespace std::literals;
+
+namespace ddwaf {
+namespace {
+
+enum class shell_flags {
+    none,
+    linux_command_opt,
+    windows_command_opt,
+};
+
+// An iterator which returns the given scalar, so that the match_iterator can be
+// used directly with a scalar without the need for a fully-fledged object iterator
+class scalar_iterator {
+public:
+    explicit scalar_iterator(const ddwaf_object *obj, const std::span<const std::string> & /*path*/,
+        const exclusion::object_set_ref & /*exclude*/, const object_limits & /*limits*/)
+        : current_(obj)
+    {}
+
+    ~scalar_iterator() = default;
+
+    scalar_iterator(const scalar_iterator &) = default;
+    scalar_iterator(scalar_iterator &&) noexcept = default;
+
+    scalar_iterator &operator=(const scalar_iterator &) = delete;
+    scalar_iterator &operator=(scalar_iterator &&) noexcept = delete;
+
+    [[nodiscard]] const ddwaf_object *operator*() { return current_; }
+    bool operator++()
+    {
+        current_ = nullptr;
+        return false;
+    }
+    [[nodiscard]] explicit operator bool() const { return current_ != nullptr; }
+    [[nodiscard]] static std::vector<std::string> get_current_path() { return {}; }
+
+protected:
+    const ddwaf_object *current_;
+};
+
+// Most shells support -c as a way to specify a shell command, however some
+// shells such as ksh allow for the first argument to be a shell command
+std::unordered_map<std::string_view, std::string_view> known_shells{
+    {"sh", "c"},
+    {"bash", "c"},
+    {"ksh", {}},
+    {"rksh", {}},
+    {"fish", "c"},
+    {"zsh", "c"},
+    {"dash", "c"},
+    {"ash", "c"},
+    {"powershell", "Command"},
+    {"powershell.exe", "Command"},
+};
+
+std::string_view basename(std::string_view path)
+{
+    auto idx = path.find_last_of(R"(\/)"sv);
```

Comment:
this is not accurate outside of windows. `a\b` is a valid filename. So it's preferable to detect if we're being compiled for windows.